### PR TITLE
Hebbian Learning in Perforant Pathway

### PR DESCRIPTION
**Pull request by abdel**
- Moved `is_hebbian_perforant` to the inside of `AHA` to fix the issue with `self` call
- Created local optimizer `LocalOptim` to facilitate manual weight updates, in a PyTorch-y way
- Created `learning_rules.py` to hold all the relevant 'local learning rules', e.g. Oja, Hebb
    - Initially, I found an implementation of Oja rule (see `OjaRule`) and utilised that as basis for the second learning rule
    - The Oja rule was adapted into `PureHebb` as the 'pure Hebbian learning' rule is described in the Ketz2013
    - The initial implementation was very inefficient especially dealing with large weights matrix (EC:CA3) and was re-written in a vectorized way which significantly increased the performance
- Extended `nn.Linear` as the `LocalConnection` class to allow us to default to untrainable weights and better control the initialisation
    - We can then instantiate both `dg_ca3` and `ec_ca3` using `LocalConnection`, and utilise the new `LocalOptim` to control weight updates for each set of connections
    - The outputs of each connection is then combined into `pc_out`
    - The weight update occurs using `pc_out` as the 'target' with respect to the corresponding connection input (e.g. `dg` or `ec_in`
- Added a new MSE metric to measure the difference between pre and post synaptic activity: `loss/dg_ca3`, `loss/ec_ca3` and `loss/pc`
- Hid the existing `PC` buffer and `PR` implementations behind the negation of the `is_hebbian_perforant` flag
- Added additional checks in `lake` to ensure it is not 'assuming' that `PR` exists
- Created new `aha_config_hebb.json` with increased study steps and removal of `PR` metrics

### CODE REVIEW

File: cls_module/cls_module/components/learning_rules.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+class LearningRule(ABC):
+  def __init__(self):
+    self.logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
+
+  def init_layers(self, model):
+    pass
+
+  @abstractmethod
+  def update(self, x, w):
```
Comment (affogato) [inline cls_module/cls_module/components/learning_rules.py]: Please add commenting on the purpose of the update rule.
Also, we should make the target an input parameter for consistency right?

File: cls_module/cls_module/components/learning_rules.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        d_w = torch.zeros(w.shape)
+        for i in range(y.shape[0]):
+            for j in range(x.shape[0]):
+                d_w[i, j] = self.c * y[i] * (x[j] - y[i] * w[i, j])
+
+        d_ws[idx] = d_w
+
+    return torch.mean(d_ws, dim=0)
+
+class PureHebbRule(LearningRule):
```
Comment (affogato) [inline cls_module/cls_module/components/learning_rules.py]: This is a variation of Oja's rule, so we need a different name, or to combine it into the Oja rule above.
I know Ketz said 'pure hebbian', but he meant 'it's not leabra with error driven'.

Comment (affogato) [reply cls_module/cls_module/components/learning_rules.py]: From Norman 2003
Hebbian Learning
The simplest form of Hebbian learning adjusts the weights in proportion to the product of the sending (x) and receiving (y) unit activations: . The weight vector is dominated by the principal eigenvector of the pairwise correlation matrix of the input, but it also grows without bound. Leabra uses essentially the same learning rule used in competitive learning or mixtures- of-Gaussians (Rumelhart & Zipser, 1986; Nowlan, 1990; Grossberg, 1976), which can be seen as a variant of the Oja normalization (Oja, 1982):

Comment (affogato) [reply cls_module/cls_module/components/learning_rules.py]: In summary, this is a variation of Oja that is used in Leabra for the Hebbian component (i.e. not error-driven)

File: cls_module/cls_module/memory/stm/aha.py [hunk 3]
```diff
@@ -231,7 +299,26 @@ def forward_pr(self, inputs, targets):
 
   def build_pc(self, pc_config, pc_input_shape):
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py:300]: I reckon we should do one major refactor where we replace all 'pc_' with 'ca3_' and 'ps_' with 'dg_'
And likewise PM, and other non biological terms. What do you think?

File: cls_module/cls_module/memory/stm/aha.py [hunk 4]
```diff
@@ -231,7 +299,26 @@ def forward_pr(self, inputs, targets):
...
+      ec_ca3 = LocalConnection(ec_size, ps_size, bias=False).to(self.device)
+      ec_ca3_optimizer = LocalOptim(ec_ca3.named_parameters(), lr=pc_config['learning_rate'])
+
+      self.add_module('ec_ca3', ec_ca3)
+      self.add_optimizer('ec_ca3', ec_ca3_optimizer)
+
+      self.learning_rule = PureHebbRule()
 
     # Initialise the buffer
     self.pc_buffer = None
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py:324]: Is there meant to be nothing really outside of `is_hebbian_perforant`?
Even the buffer is not 'built'. Is it because it is not a torch component?

File: lake/oneshot_cls.py [hunk 5]
```diff
@@ -259,19 +259,20 @@ def main():
...
-                              secondary_labels=model.features['recall']['labels'],
-                              comparison_type='match_mse')
-
-      # oneshot_metrics.compare(prefix='pr_rf_',
-      #                         primary_features=model.features['recall']['stm_pr'],
-      #                         primary_labels=model.features['recall']['labels'],
-      #                         secondary_features=model.features['study']['stm_pr'],
-      #                         secondary_labels=model.features['study']['labels'],
-      #                         comparison_type='match_mse')
+      if 'stm_pr' in model.features['study']:
```
Comment (affogato) [inline lake/oneshot_cls.py:309]: what metrics do we want if hebb ?

File: cls_module/cls_module/memory/stm/aha.py [hunk 6]
```diff
@@ -315,23 +402,60 @@ def forward_memory(self, inputs, targets, labels):
...
-    # Compute PR Mismatch
-    pr_out = outputs['pr']['pr_out']
-    pr_batch_size = pr_out.shape[0]
-    losses['pr_mismatch'] = torch.sum(torch.abs(pr_targets - pr_out)) / pr_batch_size
+        ec_ca3_in = torch.flatten(inputs, 1)
+        pre_ec_ca3_out = self.ec_ca3(ec_ca3_in)
 
-    pc_cue = outputs['ps'] if self.training else outputs['pr']['z_cue']
-    outputs['pc'] = self.forward_pc(inputs=pc_cue)
+        pre_pc_out = pre_dg_ca3_out + pre_ec_ca3_out
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py]: Is this a confusing variable name?
i.e. Is this the input to PC (including contributions from dg and from ec)

File: cls_module/cls_module/memory/stm/aha.py [hunk 7]
```diff
@@ -315,23 +402,60 @@ def forward_memory(self, inputs, targets, labels):
...
+          d_dg_ca3 = self.learning_rule.update(dg_ca3_in, pre_pc_out, self.dg_ca3.weight)
+          d_dg_ca3 = d_dg_ca3.view(*self.dg_ca3.weight.size())
+          self.dg_ca3_optimizer.local_step(d_dg_ca3)
+
+          # Update EC:CA3 with respect to ec_ca3_in (i.e. inputs)
+          d_ec_ca3 = self.learning_rule.update(ec_ca3_in, pre_pc_out, self.ec_ca3.weight)
+          d_ec_ca3 = d_ec_ca3.view(*self.ec_ca3.weight.size())
+          self.ec_ca3_optimizer.local_step(d_ec_ca3)
+
+        post_dg_ca3_out = self.dg_ca3(dg_ca3_in)
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py:438]: Just confirming - is it just to compute loss?

File: cls_module/cls_module/components/local_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
+from torch.optim.optimizer import Optimizer, required
+
+class LocalOptim(Optimizer):
+  def __init__(self, named_params, lr=required):
+    self.param_names, params = zip(*named_params)
+
+    if lr is not required and lr < 0.0:
```
Comment (affogato) [inline cls_module/cls_module/components/local_optimizer.py:7]: Looks like this should be:
`if lr is required and lr < 0.0`
i.e. if if it is not required, then it wouldn't matter if it was below 0

Comment (affogato) [reply cls_module/cls_module/components/local_optimizer.py:7]: or rather, since we know that we DO need lr, should it instead of an OR instead of an AND between these two

File: cls_module/cls_module/memory/stm/aha.py (hunk 4 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:324]: Yeah, because the `buffer` is really initialised when the first inputs come in. Aside from that, there are no layers to build and nothing else to initialise.

I guess technically, rather than `self.pc_buffer = None`, you could initialise it with 'zeros'.. but then that will just get replaced when the first step hits `forward_pc`.

File: cls_module/cls_module/components/local_optimizer.py (hunk 8 above)
Comment (abdel) [reply cls_module/cls_module/components/local_optimizer.py:7]: The conditional is checking that variable `lr` is NOT defaulting to the default value of `required`, which is how PyTorch deals with 'required' fields.

This is essentially saying: If an `lr` was passed, and its value is `< 0` then, trigger the error.

If nothing was passed, PyTorch automatically triggers the validation of the required field.

**Comment by affogato** [comment]
But it _is_ required

**Comment by affogato** [comment]
i see, it would have the actual value 'required'.
don't fully understand how that works, but it's clear that it wasn't a bug

On Wed, 14 Jul 2021 at 17:41, Gideon Kowadlo ***@***.***> wrote:

> But it _is_ required
>
>
>

-- 
Gideon Kowadlo
http://gideon.kowadlo.net

File: cls_module/cls_module/components/learning_rules.py (hunk 1 above)
Comment (abdel) [reply cls_module/cls_module/components/learning_rules.py]: done

File: cls_module/cls_module/components/learning_rules.py (hunk 2 above)
Comment (abdel) [reply cls_module/cls_module/components/learning_rules.py]: done, as discussed. Removed other rules, renamed to `OjaLeabra`

File: cls_module/cls_module/memory/stm/aha.py (hunk 6 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py]: This was resolved. Renamed to `pc_cue` and restored the `pc_buffer` functionality in the hebbian pathway

File: cls_module/cls_module/memory/stm/aha.py (hunk 7 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:438]: I added a comment to clarify this. We only run it again to compute the loss between pre and post update activity
